In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [4]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv(
    Path('Resources/lending_data.csv')   
)

# Review the DataFrame
lending_data_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [5]:
# Separate the data into labels and features

# Separate the y variable, the labels
y= lending_data_df["loan_status"]


# Separate the X variable, the features
x = lending_data_df.drop(columns="loan_status")

In [6]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [7]:
# Review the X variable DataFrame
x[:5]

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [9]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [10]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(x_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [11]:
# Make a prediction using the testing data
testing_predictions = logistic_regression_model.predict(x_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [12]:
# Generate a confusion matrix for the model
test_matrix = confusion_matrix(y_test, testing_predictions)
# Print the confusion matrix for the testing data
print(test_matrix)

[[18663   102]
 [   56   563]]


In [13]:
# Print the classification report for the model
# Create and save the training classification report
testing_report = classification_report(y_test, testing_predictions)

# Print the training classification report
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The purpose of this analysis was to evaluate the credit assessing model that was created. This model was trained using historical lending data (lending_data CSV). The factors analysed for each loan candidate were loan size, interest rate, borrower income, the ratio of debt to income, the number of accounts, derogatory marks and total debt. The target to be predicted by the model was loan_status which was either a 0 (healthy loan) or a 1( high risk for default loan). A financial institution would want to use this model to quickly predict if a new customer would be a suitable candidate for a loan at their bank, as they would not want to give out high risk loans that they would not seen a return on.The confusion matrix and classification report reveal how well the model was able to predict if customers were good or high-risk loan candidates by doing calculations on true and false positives and negatives. When looking at the classification report, three main factors were analysed:

•	Accuracy- the accuracy of the model was 0.99 meaning that the ratio of correctly predicted loan outcomes to total outcomes was 0.99. This is a high accuracy rate, however as there are many more healthy loan predictions than high risk loan predictions this alone is not a good enough metric to evaluate the model as a whole. 

•	Recall- The recall for healthy loan status (0) was 0.99 which means 99% of healthy loans were detected out of all the observations. For high-risk loan status (1) the recall was lower at 0.91 meaning out of all the observations only 0.91% of high risk loans were detected. As the recall for healthy loans was so high there is low risk of the model incorrectly predicting false negatives (high risk)

•	Precision- The precision value for healthy loan status was 1 meaning that there was a 1:1 ratio of correctly predicted healthy status to overall predicted healthy status, this means the false positive rate (false healthy status) for this model was negligible. For high-risk loans the precision rate was 0.85 meaning that the model only correctly predicted 0.85% of the overall predicted high-risk loans, and 0.15% were predicted incorrectly. 

In summary, I would not recommend that a financial institution use this model when assessing loan candidates. Whilst the model was very accurate (0.99) the accuracy score would have been outbalanced by the much higher number of overall healthy predictions (19226) to high-risk predictions (158) as this metric takes all observations into account. The model was good at predicting the healthy loans with a recall of 0.99 and precision of 1 however for the high risk loans the recall was 0.91 and the precision was 0.85, meaning only 91% of the high-risk loans were detected by the model and only 85% of the high-risk loans that were predicted were correct. For this to be used in a real-life scenario you would want precision and recall scores in the high 90’s for both the low and high-risk loans. The bank would end up with a large number of defaulting loans if it used this model. 


---